src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -247,11 +245,17 @@ public async Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toB
                     }
 
                     // Before returning the items, make sure they are sorted.
-                    provenHeadersOutput = AddToSortedDictionary(headersInCache, provenHeadersOutput);
                     provenHeadersOutput = AddToSortedDictionary(repositoryHeaders, provenHeadersOutput);
                 }
             }
 
+            if (!IsInSequence(provenHeadersOutput.ToList()))
```

Comment:
just pass the dictionary in to the method, you are enumerating like 3 times at least on this collection.